src/deluge/gui/views/performance_session_view.cpp

```diff
@@ -0,0 +1,2045 @@
+
+//creates "8 - 1 row # tags within a "row" tag"
+//limiting # of rows to the # of rows on the grid (8 = kDisplayHeight)
+void PerformanceSessionView::writeDefaultFXRowValuesToFile(int32_t xDisplay) {
+	//<row>
+	storageManager.writeOpeningTagBeginning(STRING_FOR_PERFORM_DEFAULTS_ROW_TAG);
+	storageManager.writeOpeningTagEnd();
+	char rowNumber[5];
+	//creates tags from row 8 down to row 1
+	for (int32_t yDisplay = kDisplayHeight - 1; yDisplay >= 0; yDisplay--) {
+		intToString(yDisplay + 1, rowNumber);
+		storageManager.writeTag(rowNumber, defaultFXValues[xDisplay][yDisplay] + kKnobPosOffset);
+
+		backupXMLDefaultFXValues[xDisplay][yDisplay] = defaultFXValues[xDisplay][yDisplay];
+	}
+	storageManager.writeClosingTag(STRING_FOR_PERFORM_DEFAULTS_ROW_TAG);
+}
+
+//for each FX column, write the held status, what row is being held, and what previous value was
+//(previous value is used to reset param after you remove the held status)
+void PerformanceSessionView::writeDefaultFXHoldStatusToFile(int32_t xDisplay) {
+	//<hold>
+	storageManager.writeOpeningTagBeginning(STRING_FOR_PERFORM_DEFAULTS_HOLD_TAG);
+	storageManager.writeOpeningTagEnd();
+
+	if (FXPress[xDisplay].padPressHeld) {
+		//<status>
+		storageManager.writeTag(STRING_FOR_PERFORM_DEFAULTS_HOLD_STATUS_TAG, STRING_FOR_ON);
+		//<row>
+		storageManager.writeTag(STRING_FOR_PERFORM_DEFAULTS_ROW_TAG, FXPress[xDisplay].yDisplay + 1);
+		//<resetValue>
+		storageManager.writeTag(STRING_FOR_PERFORM_DEFAULTS_HOLD_RESETVALUE_TAG,
+		                        FXPress[xDisplay].previousKnobPosition + kKnobPosOffset);
+
+		backupXMLDefaultFXPress[xDisplay].padPressHeld = FXPress[xDisplay].padPressHeld;
+		backupXMLDefaultFXPress[xDisplay].yDisplay = FXPress[xDisplay].yDisplay;
+		backupXMLDefaultFXPress[xDisplay].previousKnobPosition = FXPress[xDisplay].previousKnobPosition;
+	}
+	else {
+		//<status>
+		storageManager.writeTag(STRING_FOR_PERFORM_DEFAULTS_HOLD_STATUS_TAG, STRING_FOR_OFF);
+		//<row>
+		storageManager.writeTag(STRING_FOR_PERFORM_DEFAULTS_ROW_TAG, kNoSelection);
+		//<resetValue>
+		storageManager.writeTag(STRING_FOR_PERFORM_DEFAULTS_HOLD_RESETVALUE_TAG, kNoSelection);
+
+		backupXMLDefaultFXPress[xDisplay].padPressHeld = false;
+		backupXMLDefaultFXPress[xDisplay].yDisplay = kNoSelection;
+		backupXMLDefaultFXPress[xDisplay].previousKnobPosition = kNoSelection;
+	}
+
+	storageManager.writeClosingTag(STRING_FOR_PERFORM_DEFAULTS_HOLD_TAG);
+}
+
+//backup current layout, load saved layout, log layout change, update change status
+void PerformanceSessionView::loadPerformanceViewLayout() {
+	char modelStackMemory[MODEL_STACK_MAX_SIZE];
+	ModelStackWithThreeMainThings* modelStack = currentSong->setupModelStackWithSongAsTimelineCounter(modelStackMemory);
+
+	backupPerformanceLayout();
+	resetPerformanceView(modelStack);
+	readDefaultsFromFile();
+	logPerformanceLayoutChange();
+	updateLayoutChangeStatus();
+}
+
+//read defaults from XML
+void PerformanceSessionView::readDefaultsFromFile() {
```

Comment:
I'm a little nervous about this inline sdcard work, but setting up the machinery to defer it seems like something that should be done in followup work if it turns out to be a problem -- since we're just loading 1 cluster from the filesystem root *hopefully* it's fairly unlikely to click.

Reply:
What do you mean by "inline sd card work"?

Do you mean like this should be moved to a save/load preset UI?

Yea if you we could defer it to follow up work that would be great.

I have on my action item list a request to load preset from a folder and to be able to save presets with different names using the keyboard grid UI.

Reply:
The `storageManager` calls end up doing I/O on the SDCard. It should be fine for an initial pass.